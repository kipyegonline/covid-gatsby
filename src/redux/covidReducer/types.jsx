export const C = {
    ADD_COVID: "ADD_COVID",
    SORT_COVID: "SORT_COVID",
    ADD_TIMESET: "ADD_TIMESET",
    ADD_COUNTRY: "ADD_COUNTRY",
    REMOVE_COUNTRY: "REMOVE_COUNTRY",
    LOADING_MESSAGE: "LOADING_MESSAGE",
    SET_MONTH: "SET_MONTH",
    BTN_CLICKED: "BTN_CLICKED",
    CLICKED_BTN: "CLICKED_BTN"
}