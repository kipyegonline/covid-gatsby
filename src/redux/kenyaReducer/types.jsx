export const C = {
    ADD_KE: "ADD_KE",
    SORT_GRAPH: "SORT_GRAPH",
    SELECTED_VALUE: "SELECTED_VALUE",
    LOADING_MESSAGE: "LOADING_MESSAGE"
}