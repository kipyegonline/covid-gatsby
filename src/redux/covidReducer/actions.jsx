import { C } from "./types"

export const addCovid = (payload) => ({
    type: C.ADD_COVID,
    payload
})
export const addTimeSeries = (data) => ({
    type: C.ADD_TIMESET,
    payload: transformData(data)
})


export const addCountry = (payload) => ({
    type: C.ADD_COUNTRY,
    payload
})

export const removeCountry = (payload) => ({
    type: C.REMOVE_COUNTRY,
    payload
})

export const loadingMessage = (payload) => ({
    type: C.LOADING_MESSAGE,
    payload
})
export const setMonth = (payload) => ({
    type: C.SET_MONTH,
    payload
})
export const sortCovid = (sort, cwd) => ({
    type: C.SORT_COVID, sort, sorted: cwd
})
export const sendBtn = (payload) => ({
    type: C.BTN_CLICKED,
    payload
})




const fetchData = async (url) => {
    try {

        const res = await fetch(url);

        if (res.status) {
            const data = await res.json();
            return data.Countries;
        }

    }
    catch (error) {
        console.log('fetch err data:', error.message)
    }


}

const fetchContinent = async (region) => {
    try {

        let global = "https://restcountries.eu/rest/v2/all"

        let url = region === "global" ? global : `https://restcountries.eu/rest/v2/region/${region}`

        const res = await fetch(url);
        const data = await res.json();

        let dataset = data.map(item => {
            if (item.name === "United States of America") {
                item.name = "US"
            }
            return {
                name: item.name,
                flag: item.flag
            }
        })
        return dataset;
    }
    catch (error) {
        console.log('fetch err', error.message)
    }

}



export function getNations(data, continent, countryKey, callback) {
    let dataholder = []

    if (data) {

        dataholder = data.filter(item => continent.find(afr => {

            if (item[countryKey] === "United States of America") {
                item[countryKey] = "US";
            }
            if (afr["name"] === item[countryKey]) {
                item.flag = afr.flag;
                return item
            }

        })





        )


        callback(dataholder)


    }
}

export async function covid19(callback, ur1, region, tracker) {

    try {
        tracker({ msg: "loading " + region, spin: true })

        const [covid, africa] = await Promise.all([fetchData(ur1), fetchContinent(region)])
        if (covid === undefined || africa === undefined) {
            tracker({
                msg: `Error loading ${region}..try again`, spin: true
            })
            setTimeout(() => tracker({ msg: "", spin: false }), 2000)
        } else {
            getNations(covid, africa, "Country", callback)
            tracker({ msg: "", spin: false })
        }

    }
    catch{
        console.log("promise errrr")
    }

}


///get time series of cases


const fetchEntry = ur1 => fetch(ur1).then(res => res.json()).then(data => data).catch(err => console.error(err.message));

export async function covid19Series(callback, ur1, region, tracker) {

    try {

        tracker({ msg: "loading data", spin: true });

        const [covidData, continent] = await Promise.all([fetchEntry(ur1), fetchContinent(region)])
        // console.log('seriesssss', covidData, continent)
        if (covidData === undefined || continent === undefined) {

            tracker({ msg: "Error loading..try again", spin: false })
            setTimeout(() => tracker({ msg: "", spin: false }), 2000)
        } else {
            tracker({ msg: "", spin: false })
            getNations(covidData, continent, "countryregion", callback, tracker)
        }

    }
    catch (error) {
        console.log("Line Error: ", error.message)
        tracker({ msg: error.message, spin: false })

    }


}


export const transformData = (dataset) => {
    if (dataset) {
        let counter = 0;


        return dataset.map(data => {
            let timeData = []
            const bigO = data.timeseries
            for (let ob in bigO) {

                const confirmed = bigO[ob].confirmed;
                const deaths = bigO[ob].deaths;
                const recovered = bigO[ob].recovered;
                const stats = {
                    date: ob,
                    confirmed,
                    deaths,
                    recovered
                }

                timeData = [...timeData, stats]
            }
            data.latestupdate = new Date(data.lastupdate.split("T")[0])
            let count = 0;
            data.stats = timeData.map(time => {
                return {
                    ...time,
                    date: new Date(time.date),
                    id: count++
                }


            })
            data.latest = timeData[timeData.length - 1].confirmed;
            data.deaths = timeData[timeData.length - 1].deaths;
            data.recovered = timeData[timeData.length - 1].recovered;
            data.date = timeData[timeData.length - 1].date;
            data.id = counter++;

            return data;
        })
            .sort((a, b) => b.latest - a.latest)
    }

}
let count = 0;
export const filterByMonth = (dataset, month) => {
    let holder = [];

    const res = dataset.map(d => {
        return d.filter(item => {
            if (item.date.getMonth() >= month) {

                return item
            }
        })
    })

    return res;
}

