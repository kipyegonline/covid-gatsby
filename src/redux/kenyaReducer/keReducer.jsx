import { C } from "./types"
const initState = {
    cwd: [],
    all: [],
    sort: true,
    message: "",
    spinner: false

}
function keRducer(state = initState, action) {
    switch (action.type) {
        case C.ADD_KE:
            return {
                ...state,
                all: action.payload,
                cwd: action.payload.find(d => d.countryregion === "Kenya") == undefined ? [] : action.payload.find(d => d.countryregion === "Kenya")
            }

            
        case C.SELECTED_VALUE:
            return {
                ...state,
                cwd: state.all.find(d => d.countryregion === action.payload)
            }
        case C.SORT_GRAPH:

            return { ...state, sort: action.sorted }

        case C.LOADING_MESSAGE:
            return {
                ...state,
                spinner: action.payload.spinner,
                message: action.payload.message
            }
        default:
            return state
            break;
    }
}
export default keRducer;




