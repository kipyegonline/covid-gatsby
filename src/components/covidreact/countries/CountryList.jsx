import styled from "styled-components";
import React, { useState, useEffect } from "react";
import { useDispatch } from "react-redux";
import * as actions from "../../../redux/covidReducer/actions";
import { formatNumLocale, paginationDevice } from "../../helpers";
import Pagination from "./pagination";

const Button = styled.button.attrs((props) => ({
  className: props.classlist,
  onClick: props.handleClick,
}))`
  max-width: 100px;
  outline: none;
  padding: 0.5rem;
`;
const ToolTip = styled.div.attrs((props) => ({
  className: "flags",
}))`
  width: ${(props) => props.data.w || 0}px;
  border-radius: 10px;
  background: ${(props) => props.data.bg};
  padding: 0.5rem;
  font-size: 1rem;
  margin: 5px;
  top: ${(props) => props.data.y || 0}px;
  left: ${(props) => props.data.x || 0}px;
  position: absolute;
  z-index: 110;
  height: auto;
  opacity: ${(props) => (props.tool ? 1 : 0)};
  transition: all 0.5s ease-in;
  box-shadow: -2px -2px 3px #fff, 2px 2px 3px #fff;
  pointer-events: none;
`;
const Para = styled.p.attrs((props) => ({ className: props.classlist }))`
  padding: 0.25rem;
  margin: 0.25rem;
  font-family: roboto;
  font-size: 1rem;
`;

const CountryList = ({
  btns,
  dataset,
  spinner,
  message,
  clickedBtn,
  addCovid,
  loadingMessage,
}) => {
  
  const dispatch = useDispatch();

  const handleClick = (data) => {
    const url =
      "https://wuhan-coronavirus-api.laeyoung.endpoint.ainize.ai/jhu-edu/timeseries";
    dispatch(actions.sendBtn(data.id));
    // (callback, ur1, region, tracker)
    actions.covid19Series(addCovid, url, data.val, loadingMessage);
  };

  const [perPage, setPage] = useState(paginationDevice());
  const [pageNos, setPageNo] = useState(0);
  const [current, setCurrent] = useState(0);
  useEffect(() => {
    //get the number of pages and set state
    let pg = Math.ceil(dataset.length / perPage);
    setPageNo(pg);
  }, []);

  const getPage = (pg) => setCurrent(pg);

  const btnfill = "btn btn-md btn-info mr-2 my-2 blue lighten-1 ";
  const btnClicked = "btn btn-md btn-info mr-2 my-2 red lighten-2 ";

  return (
    <div className="row">
      <div className="col-md-12 col-lg-9 col-sm-12">
        <p className="my-2 p-1">Mouse over the flags to view covid-19 report</p>
        <div className="my-2 flag-btns">
          {btns.map((continent) => (
            <Button
              key={continent.id}
              classlist={continent.selected ? btnClicked : btnfill}
              handleClick={() => handleClick(continent)}
            >
              {continent.continent}
            </Button>
          ))}
          {spinner ? (
            <p className="text-success">Loading {clickedBtn.toUpperCase()}</p>
          ) : (
            <p className="text-danger">{message}</p>
          )}
        </div>
        <div className="main-flags">
          {dataset
            .slice(current * perPage, current * perPage + perPage)
            .map((data, i) => (
              <CountrySet data={data} key={data.id} />
            ))}
        </div>
        <Pagination pages={pageNos} setPage={getPage} current={current} />
      </div>

      <div className="col-md-12 col-lg-3 col-sm-12 mt-5 table-responsive">
        <h5 className="text-center">COVID-19 summary</h5>
        <CountryTable
          dataset={dataset.slice(
            current * perPage,
            current * perPage + perPage
          )}
          current={current * perPage}
        />
      </div>
    </div>
  );
};
export default CountryList;

const CountrySet = ({ data }) => {
  const [hover, setHover] = useState(false);
  const [info, setInfo] = useState({});

  const handleMouseEnter = (e, data) => {
    const { left, top } = e.target.getBoundingClientRect();
    setInfo({
      x: window.screenX,
      y: window.screenY + 50,
      w: 150,
      bg: "#fff",
      deaths: data.deaths,
      recovered: data.recovered,
      latest: data.latest,
      countryregion: data.countryregion,
      flag: data.flag,
    });

    setHover(true);
  };
  const handeLeave = () => {
    setHover(false);
  };

  return (
    <div className="card card-flag mx-2">
      <ToolBar hover={hover} data={info} />
      <h5 className="card-title text-center">{data.countryregion}</h5>
      <img
        className="card-img-top"
        onMouseOver={(e) => handleMouseEnter(e, data)}
        onMouseLeave={handeLeave}
        src={data.flag}
        alt={data.countryregion}
      />
    </div>
  );
};
const CountryTable = ({ dataset, current }) => (
  <table className="table table-hover table-bordered">
    <thead>
      <tr>
        <th>#</th>
        <th>Country</th>
        <th>Confirmed</th>
        <th>Deaths</th>
        <th>Recovered</th>
      </tr>
    </thead>
    <tbody>
      {dataset.map((data, i) => (
        <tr key={i}>
          <td>{i + current + 1}</td>
          <td>{data.countryregion}</td>
          <td className="text-info">{formatNumLocale(data.latest)}</td>
          <td className="text-danger">{formatNumLocale(data.deaths)}</td>
          <td className="text-success">{formatNumLocale(data.recovered)}</td>
        </tr>
      ))}
    </tbody>
  </table>
);

const ToolBar = ({ hover, data }) => (
  <ToolTip tool={hover} data={data}>
    <Para claslist="text-center">{data.countryregion}</Para>
    <Para classlist="text-info">
      Confirmed: {""}
      {formatNumLocale(data.latest)}
    </Para>
    <Para classlist="text-danger">
      Deaths: {""}
      {formatNumLocale(data.deaths)}
    </Para>
    <Para classlist="text-success">
      Recovered: {""}
      {formatNumLocale(data.recovered)}
    </Para>
  </ToolTip>
);
