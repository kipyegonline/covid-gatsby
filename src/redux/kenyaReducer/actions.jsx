import { C } from "./types"
import { transformData } from "../covidReducer/actions"
export const addKe = (data) => ({ type: C.ADD_KE, payload: transformData(data) })

export const sendValue = value => ({ type: C.SELECTED_VALUE, payload: value })
export const loadingMessage = value => ({ type: C.LOADING_MESSAGE, payload: value })
export const fetchKe = async (url) => {
    try {

        const res = await fetch(url)
        if (res.status) {
            const data = await res.json()

            return transformData(data);
        }

    }
    catch (err) {
        // console.error("Ke fetch", err.message)
    }


}