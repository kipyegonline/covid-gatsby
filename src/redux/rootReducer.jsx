import { combineReducers } from "redux"
import covidReducer from "./covidReducer/covidReducer"
import keReducer from "./kenyaReducer/keReducer"
export default combineReducers({ covid: covidReducer, ke: keReducer })