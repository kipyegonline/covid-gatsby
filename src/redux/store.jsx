import React from "react"
import { createStore } from "redux"
import { Provider } from "react-redux"
import rootReducer from "./rootReducer"

const devtools = () => {
  if (globalThis.window) {
    return window.__REDUX_DEVTOOLS_EXTENSION__
      ? window.__REDUX_DEVTOOLS_EXTENSION__()
      : f => f
  }
}
export const store = createStore(rootReducer, devtools())

const ReduxWrapper = ({ element }) => (
  <Provider store={store}>{element}</Provider>
)
export default ReduxWrapper
