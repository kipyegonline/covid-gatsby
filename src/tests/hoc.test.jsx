import React from "react";
import Enzyme, { mount, shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import Adapter from "enzyme-adapter-react-16";
import { createStore } from "redux";
import CountryList from "../components/covidreact/countries/CountryList";
Enzyme.configure({ adapter: new Adapter() });

import Expandable, { MyForm } from "./hoc";
import keReducer from "../redux/kenyaReducer/keReducer";
import rootReducer from "../redux/rootReducer";
import { C } from "../redux/kenyaReducer/types";
import { addKe } from "../redux/kenyaReducer/actions";
import LineChart from "../components/covidreact/line/lineChart";
import Line from "../components/covidreact/line/line";

jest.mock(
  "./mock.js",
  () => ({ rating, colorId, onRate = (f) => f, onRemove = (f) => f }) => (
    <div>
      <button className="rate" onClick={() => onRate(rating)} />
      <button className="remove" onClick={() => onRemove(colorId)} />
    </div>
  )
);

describe("<ColorList/> UI component", () => {
  describe("Rating a color", () => {
    let _rate = jest.fn(),
      _remove = jest.fn();

    beforeEach(() => {
      mount(
        <ColorList
          colors={[
            { color: "red", id: 1, title: "rebeccapurple" },
            { color: "green", id: 2, title: "pink" },
          ]}
          onRate={_rate}
        />
      )
        .find("button.rate")
        .first()
        .simulate("click");
      mount(
        <ColorList
          colors={[
            { color: "red", id: 1, title: "rebeccapurple" },
            { color: "green", id: 2, title: "pink" },
          ]}
          onRemove={_remove}
        />
      )
        .find("button.remove")
        .simulate("click");
    });
    it.skip("invokes onRate handler", () => expect(_rate).toBeCalledTimes(1));
    it.skip("removes the correct Color", () =>
      expect(_remove).toBeCalledWith(1));
  });
});

//unit tests for another component
describe("Expands HOC", () => {
  let props,
    wrapper,
    ComposedComponent,
    MockComponent = ({ collapsed, expandCollapse }) => (
      <div onClick={expandCollapse}>{collapsed ? "collapsed" : "expanded"}</div>
    );

  describe("rendering UI", () => {
    beforeAll(() => {
      ComposedComponent = Expandable(MockComponent);
      wrapper = mount(
        <ComposedComponent hidden={true} foo="foo" gnar="gnar" />
      );
      props = wrapper.find(MockComponent).props();
    });

    test("starts off collapsed", () => expect(props.collapsed).toBe(true));
    test("passes the expand collapse func to composed", () =>
      expect(typeof props.expandCollapse).toBe("function"));
    test("passes additional gnar prop to composed", () =>
      expect(props.gnar).toBe("gnar"));
    test("hidden prop", () => expect(props.hidden).toBe(true));
  });

  describe("Expand Collapse Functionality", () => {
    let instance;
    beforeAll(() => {
      ComposedComponent = Expandable(MockComponent);
      wrapper = mount(<ComposedComponent collapsed={false} />);
      instance = wrapper.instance();
      wrapper.update();
    });
    test.skip("renders the mock component as root element", () =>
      expect(wrapper.find(MockComponent)).toBe(MockComponent));
    test("starts off expanded", () =>
      expect(instance.state.collapsed).toBe(false));

    test.skip("toggles the collapsed state", () =>
      expect(instance.state.collapsed).toBe(true));
  });
});
let pp1 = [
  { name: "Vince", time: new Date().toDateString() },
  { name: "Jules", time: new Date().toDateString() },
  { name: "Mark", time: new Date().toDateString() },
];

describe("Ke  reducer", () => {
  const action1 = { type: C.ADD_KE, payload: pp1 };
  const res = keReducer([], action1);
  test("add cwd", () => expect(res).toEqual({ cwd: [], all: pp1 }));

  const act1 = { type: C.SORT_GRAPH, sorted: true };
  test("sorting", () => expect(keReducer({}, act1)).toEqual({ sort: true }));
});

describe("<MyForm> component", () => {
  let wrapper, _submit, _change, _click;

  describe("Main Wrapper", () => {
    beforeEach(() => {
      _click = jest.fn();
      _submit = jest.fn();
      _change = jest.fn();
      wrapper = shallow(
        <MyForm submit={_submit} onChange={_change} onClick={_click} />
      );
    });
    //afterEach(() => wrapper.unmount());

    test("snapshots", () => expect(wrapper.html()).toMatchSnapshot());
    it("contains a list of three items", () => {
      expect(wrapper.find("li")).toHaveLength(3);
      expect(
        wrapper
          .find("li")
          .first()
          .text()
      ).toEqual("All you need to know X");
      expect(wrapper.find("h4").text()).toHaveLength(0);
    });
    it("Text area changes", () => {
      wrapper.find("textarea").simulate("change", {
        target: { value: "Ava Max is sweeter but  a psycho...." },
      });

      expect(wrapper.find("input[type='submit']").hasClass("clout")).toBe(true);
      expect(wrapper.find("textarea").prop("value")).toEqual(
        "Ava Max is sweeter but  a psycho...."
      );
      expect(wrapper.find("h4").text()).toHaveLength(36);
      wrapper.find("form").simulate("submit");
      wrapper.update();

      expect(wrapper.find("textarea").prop("value").length).toBe(0);
      expect(wrapper.find("li").length).toEqual(4);
    });

    test("when an item is deleted", () => {
      wrapper
        .find("span")
        .first()
        .simulate("click");
      expect(wrapper.find("li").length).toEqual(2);
    });
    test("submit button", () => {
      expect(wrapper.find('input[type="submit"]').props()).toEqual({
        type: "submit",
        className: "btn blue lighten-1",
        value: "Send Naah",
      });
    });
  });
});

describe("<Events/>", () => {
  let _change = jest.fn();
  let wrapper = shallow(<MyForm formTitle="My Form" onChange={_change} />);
  wrapper
    .find("textarea")
    .simulate("change", { target: { value: "Tenerife sea" } });
  test("snapshot match", () => expect(toJSON(wrapper)).toMatchSnapshot());
  test("change event", () => {
    const wrapped = shallow(<MyForm formTitle="My Form" onChange={_change} />);
    expect(wrapped.find("h3").text()).toEqual("My Form");
    expect(wrapped.find("p").text()).toEqual("You just want attention");
  });
});

/**
 * Jest config
 * {
 * "setUpFiles":[
 * "raf/polyfill",
 * "<rootDir>/src/tests/setupTest.js"],
 * "snapshotSerializers":["enzyme-to-jsom/serializer"]}
 */
