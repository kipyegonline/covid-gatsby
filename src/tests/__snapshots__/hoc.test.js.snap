// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Events/> snapshot match 1`] = `
<form
  className="form p-5 mx-auto my-2 gray lighten-2"
  onSubmit={[Function]}
>
  <h3>
    My Form
  </h3>
  <h4>
    Tenerife sea
  </h4>
  <p>
    You just want attention
  </p>
  <ul
    className="list-group"
  >
    <li
      className="list-group-item"
      key="1"
    >
      All you need to know
       
      <span
        className="text-danger ml-5"
        onClick={[Function]}
      >
        X
      </span>
    </li>
    <li
      className="list-group-item"
      key="2"
    >
      I need your love
       
      <span
        className="text-danger ml-5"
        onClick={[Function]}
      >
        X
      </span>
    </li>
    <li
      className="list-group-item"
      key="3"
    >
      Unit testing manenos
       
      <span
        className="text-danger ml-5"
        onClick={[Function]}
      >
        X
      </span>
    </li>
  </ul>
  <label>
    Enter text
  </label>
  <textarea
    className="form-control my-2"
    onChange={[Function]}
    value="Tenerife sea"
  />
  <input
    className="btn blue lighten-2 clout"
    type="submit"
    value="Send Naah"
  />
</form>
`;

exports[`<MyForm> component Main Wrapper snapshots 1`] = `"<form class=\\"form p-5 mx-auto my-2 gray lighten-2\\"><h3></h3><h4></h4><p>You just want attention</p><ul class=\\"list-group\\"><li class=\\"list-group-item\\">All you need to know <span class=\\"text-danger ml-5\\">X</span></li><li class=\\"list-group-item\\">I need your love <span class=\\"text-danger ml-5\\">X</span></li><li class=\\"list-group-item\\">Unit testing manenos <span class=\\"text-danger ml-5\\">X</span></li></ul><label>Enter text</label><textarea class=\\"form-control my-2\\"></textarea><input type=\\"submit\\" class=\\"btn blue lighten-1\\" value=\\"Send Naah\\"/></form>"`;
